### ISSUE TITLE
[dynamic][inductor][recompilations] Static guard on tensor dim size

### ISSUE THREAD

**Issue by anijain2305** [pytorch/pytorch#139936]
### 🐛 Describe the bug

~~~

import torch


def fn(x):
    return x.chunk(4)

opt_fn = torch.compile(fn, backend="inductor", fullgraph=True)


for i in range(2,12):
    x = torch.randn(4*i)
    fn(x)
    opt_fn(x)
~~~

With "eager", there is only 1 recompile. With inductor, we keep on recompiling.

Also affects TorchTune chunking.

### Versions

NA

cc @ezyang @chauhang @penguinwu @bobrenjc93 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @aakhundov

**Comment by bdhirsh** [comment]
specialization is happening here https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L1674

**Comment by ezyang** [comment]
TODO is relevant
